File: cpp/3.cpp
#include <bits/stdc++.h>
using namespace std;

#define ul unsigned long long int

int main() {
  ul num = 600851475143, maxPrimeFactor;

  if (num % 2 == 0) {
    maxPrimeFactor = 2;
    while (num % 2 == 0) num /= 2;
  }

  for (int i = 3; i <= sqrt(num); i += 2)
    if (num % i == 0) {
      maxPrimeFactor = i;
      while (num % i == 0) num /= i;
    }

  if (num > 2) maxPrimeFactor = num;

  printf("%lld\n", maxPrimeFactor);
}

File: py/3.py
from math import sqrt

num = 600851475143
max_prime_factor = 0

if num % 2 == 0:
    max_prime_factor = 2
    while num % 2 == 0:
        num //= 2

for i in range(3, int(sqrt(num)) + 1, 2):
    if num % i == 0:
        max_prime_factor = i
        while num % i == 0:
            num //= i

if num > 2:
    max_prime_factor = num

print(max_prime_factor)
